### DISCUSSION TITLE
Is there any way to transfer Autogluon into ONNX for deploying in C++ environment

### DISCUSSION THREAD

**Discussion by Jeremylee1234** [autogluon/autogluon#2972] [Q&A] [answered]
Hi! I'm currently working in a project of transferring autogluon project to ONNX for further deployment including feature generators and the ensemble models. 
By far, I have successfully transfer the Text_gram_special_generator to onnx, but for other feature generators, like the text_special one is about count the string length etc, and Onnx doesn't release such operator to address it. Can you suggest anyway for this issue? 
And for transferring ensemble model into onnx, how to transfer the ensemble model to ONNX, are there any documents or posts could be useful for this issuse? 

Thanks!

**Comment by liangfu** [comment, marked as answer]
Thanks for your attention. I have worked on converting the tabular models to onnx. For now, we are able to support model prediction for RandomForest and ExtraTrees model. I also worked on converting the ColumnTransformer in sklearn to onnx via skl2onnx, and this has been integrated with TabularNN model. To use this features, please refer to

https://auto.gluon.ai/stable/tutorials/tabular_prediction/tabular-deployment.html#compile-models-for-maximized-inference-speed

For transforming the whole ensemble model to ONNX, I think the key blockers are
1. the conversion from XGBoost, LightGBM, and CatBoost model to onnx isn't always accurate
2. ONNX has a limited set of operators. implementing some operators (e.g. searchsorted) in onnx takes a lot of risky hack that isn't very much reliable.

For supporting autogluon-multimodal with onnx, there is a open issue to track the progress: https://github.com/autogluon/autogluon/issues/2895 Please stay tuned for future updates.

**Reply by Jeremylee1234** [reply]
Thank you LiangFu, 
I have convtered bagged LightGBM model to ONNX, and there is a issue that one of the child LightGBM model outputs with onnx is highly different from the ouputs from autogluon which might hugely affects the accuracy. 

Also, as you mentioned that there are limited ONNX operators to use, is there any other way for deploying the text_special_generator in C++ environment?
Best Regards,

**Reply by Innixma** [reply]
@Jeremylee1234 Thanks for your efforts! Would you be interested in sharing the code you are using to transform LightGBM to ONNX? This would be of interest to us as we may eventually want to naturally support LightGBM compilation, and your code might be a good reference.

Additionally, for converting to ONNX for the preprocessing, you could potentially just disable that feature generator (TextSpecial) if it doesn't cost you too much accuracy to do so. Also, if you are able to somehow preprocess the data before sending to AutoGluon, you could instead set `feature_generator=None` and avoid having to worry about the preprocessing within AutoGluon, but that has its own challenges. Curious to know if you do end up finding a solution, this kind of question has been asked a lot!

**Reply by Jeremylee1234** [reply]
OK, I'll share the code next Monday

**Reply by Jeremylee1234** [reply]
[(url)](https://github.com/Jeremylee1234/Autogluon2onnx.git)
Here is the repository for transfer the LightGBM model to onnx. And it is pretty much the  first time I works on a open source project. Plz don't judge me if the code doesn't meet the standard...... lol

And right now, I'm working on transfering the NeuralTorch to onnx, like @liangfu said, there is a sklearn ColumnTransformer preprocessing function in the predict method and the issue is that ColumnTransformer contains a QuantileTransformer which skl2onnx library doesn't contain the related operator by far, I'm curious how you address this issue.

I've looked at skl2onnx repository and their maintainer tried to implement the QuantileTransformer ONNX converter, but it looks like it failed ultimately somehow.
[https://github.com/onnx/sklearn-onnx/pull/705](url)

**Reply by liangfu** [reply]
> there is a sklearn ColumnTransformer preprocessing function in the predict method and the issue is that ColumnTransformer contains a QuantileTransformer which skl2onnx library doesn't contain the related operator by far, I'm curious how you address this issue.

The [quantile_transformer_converter in onnx.py](https://github.com/autogluon/autogluon/blob/master/tabular/src/autogluon/tabular/models/tabular_nn/compilers/onnx.py) would implement such conversion.

**Reply by Jeremylee1234** [reply]
> The [quantile_transformer_converter in onnx.py](https://github.com/autogluon/autogluon/blob/master/tabular/src/autogluon/tabular/models/tabular_nn/compilers/onnx.py) would implement such conversion.

@liangfu Thanks liangfu,
I found out that the outputs of the ColumnTransformer is slightly different from the outputs from the onnx converted, which makes classification results of onnx model doesn't match the results from the predictor 
![image](https://user-images.githubusercontent.com/42887249/225535882-2909cb30-f2c6-4c5c-b619-53ac2948caab.png)

**Reply by Jeremylee1234** [reply]
@liangfu @Innixma 
For the bagged model, what I'm thinking about is about adding mean operator and argamx operator to get the final result. Which would be like the onnx graph below. Am I on the right track? 
![image](https://user-images.githubusercontent.com/42887249/228116000-39db48dd-117e-4e13-885e-738899dffcd4.png)

**Reply by liangfu** [reply]
I think you're on the righ track, but notice that argmax might not work for problem_types other than classification.
https://github.com/autogluon/autogluon/blob/bdf0c66a6c432b54136c59c3d08416fe8a665464/core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py#L320-L334

**Reply by Jeremylee1234** [reply]
Got it! Thanks.

**Comment by rxjx** [comment]
Is there a way to convert CLIP multimodal models to ONNX as well? The documentation lists export_onnx as part of the multimodal API but when I try to export a model for image_text_similarity I get "NotImplementedError: export_onnx doesn't support model type <class 'NoneType'>". I see the following line in the source:
supported_models = (TimmAutoModelForImagePrediction, HFAutoModelForTextPrediction, MultimodalFusionMLP)

**Reply by Innixma** [reply]
@rxjx I'd recommend searching if there is a GitHub issue on this. If not, feel free to create one so we can track it. Definitely if the API isn't working as described, we would want to take a look to improve the functionaliy.

**Reply by rxjx** [reply]
I didn't find an issue for this. I'll create one. Thanks.

**Reply by liangfu** [reply]
Thanks for filing the issue. This is part of the effort in supporting export_onnx for all model, which is tracked via https://github.com/autogluon/autogluon/issues/2895 .

**Comment by jmren168** [comment]
> Hi! I'm currently working in a project of transferring autogluon project to ONNX for further deployment including feature generators and the ensemble models. By far, I have successfully transfer the Text_gram_special_generator to onnx, but for other feature generators, like the text_special one is about count the string length etc, and Onnx doesn't release such operator to address it. Can you suggest anyway for this issue? And for transferring ensemble model into onnx, how to transfer the ensemble model to ONNX, are there any documents or posts could be useful for this issuse?
> 
> Thanks!

Hi @Jeremylee1234 ,

I'm also trying to transfer feature generators to ONNX format. Do you have any guides or suggesstions? Any helps are appreciated.

**Reply by Jeremylee1234** [reply]
what types of feature generators are you working on?
Since most of feature generators of autogluon are using sklearn, [https://github.com/onnx/sklearn-onnx](url)
this repository has done a lot of works on converting sklearn models and preprocessing ways to onnx.

**Reply by jmren168** [reply]
Hi @Jeremylee1234,

I'm working on re-writing existed feature generators, e.g., [fillna.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/fillna.py). Our team hope all the built-in feature generators can be transferred to onnx format, becuase we need to run AutoGluon in C/C++ environment (without pkl files). Have you ever tried to re-write an existed feature generator? Any guides are good for us, and thanks again.

**Reply by jmren168** [reply]
Hi @Jeremylee1234,

Could you run onnx models (e.g., SVM or RF) **without pickle files** in C++ environment after executing AutoGluon to generate onnx models? I ask this because if I delete pickle files of models, then AutoGluon cannot predict in Python Environment only using onnx models. Thanks.

**Reply by Jeremylee1234** [reply]
Yeah, I did multiple autogluon models to onnx models. Like lightgbm,RF,Catboost, etc. But only for classification and regression tasks. And only works for float input. 

The workflow I'm doing is after you trained on Autogluon, you need to use python pickle library to load the saved Autogluon predictor and model files to get the corresponding data preprocessors and models. Then, writing python scripts to transfer different model to onnx file, the skl2onnx package provides functions to transfer the model object to onnx graph.

**Reply by Jeremylee1234** [reply]
fillna would be better for implementation in C/C++ since ONNX doesn't provide such operators for doing that.

**Comment by jmren168** [comment]
> Yeah, I did multiple autogluon models to onnx models. Like lightgbm,RF,Catboost, etc. But only for classification and regression tasks. And only works for float input.
> 
> The workflow I'm doing is after you trained on Autogluon, you need to use python pickle library to load the saved Autogluon predictor and model files to get the corresponding data preprocessors and models. Then, writing python scripts to transfer different model to onnx file, the skl2onnx package provides functions to transfer the model object to onnx graph.

@Jeremylee1234 Thanks for the quick reply. 

"Then, writing python scripts to transfer **different model** to onnx file", does **different model** here include both data preprocessors (e.g., text_gram) and models (e.g., rf)? Or only models (e.g., rf)?

I ask this because fillna is also included in data preprocessors, and after training on Autogluon, the pickle files should include fillna. If so, do you transfer fillna to onnx file or implementit it in C/C++? 

BTW, could you please tell us in the following feature generators, which one do you successfully transfer to onnx file? 
We're trying to transfer because these generators are executed in training a AutoGluon tabular predictor.
[astype.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/astype.py)
[fillna.py ](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/fillna.py)
[identity.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/identity.py)
[category.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/category.py)
[memory_minimize.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/memory_minimize.py)
[datetime.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/datetime.py)
[text_special.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/text_special.py)
[text_ngram.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/text_ngram.py)
[isnan.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/isnan.py)
[drop_unique.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/drop_unique.py)
[drop_duplicates.py](https://github.com/autogluon/autogluon/blob/master/features/src/autogluon/features/generators/drop_duplicates.py)

**Reply by Jeremylee1234** [reply]
In our case, all trained Autogluon models don't execute these feature generators. I did text_ngram.py generator to onnx. And for text_special.py it just did some regex which could be easily implemented in your C++ environment. 
"Then, writing python scripts to transfer different model to onnx file", does different model here include both data preprocessors (e.g., text_gram) and models (e.g., rf)? Or only models (e.g., rf)? I only include models to onnx which is the most complicated part since different model type supposed to use different package to transfer to onnx file. And if your models are weighted ensemble, bagging and stacked types which will need more works to do. I suggest you to implement these generators in your C++ code, and transfer the model files to onnx file and using onnxruntime to run in C++ environment. 
Also, I suggest you to look at autogluon API documents, and using debug mode to running your model to see how the source code works.

**Reply by jmren168** [reply]
@Jeremylee1234
Thanks for the explanation and I appreciate your kindly help.

**Comment by LukaszCmielowski** [comment]
The thread is few years old but I wonder if there are any updates in that area ?. From enterprises perspective having trained model in the ONNX format is very important (reduced the burden of runtime and model lifecycle).
